src/mame/misc/anes.cpp

```diff
@@ -38,149 +38,587 @@ namespace {
+	m_palette->set_pen_color(offset, rgb_t(pal5bit(dat >> 0), pal5bit(dat >> 5), pal5bit(dat >> 10)));
+}
+
+void anes_state::palette_offset_w(uint8_t data)
+{
+	if (m_palette_enable != 0x01)
+	{
+		logerror("write to palette_offset_w when not enabled %02x (enable %02x)\n", data, m_palette_enable);
+	}
+	else
+	{
+		m_palette_offset = data;
+	}
+}
+
+void anes_state::palette_offset_msb_w(uint8_t data)
+{
+	if (m_palette_enable != 0x01)
+	{
+		logerror("%s: write to palette_offset_msb_w when not enabled %02x (enable %02x)\n", machine().describe_context(), data, m_palette_enable);
+	}
+	else
+	{
+		m_palette_offset_msb = data;
+	}
+}
+
+void anes_state::palette_active_bank_w(uint8_t data)
+{
+	m_palette_active_bank = data & 0xf;
+}
+
+uint8_t anes_state::palette_data2_r()
+{
+	if (m_palette_enable != 0x01)
+	{
+		logerror("read from palette_data2_r when not enabled\n");
+	}
+	else
+	{
+		int offs = (m_palette_offset + (m_palette_offset_msb << 8)) & 0x1fff;
+		return palette_data_lsb[offs];
+	}
+	return 0x00;
+}
+
+
+void anes_state::palette_data2_w(uint8_t data)
+{
+	if (m_palette_enable != 0x01)
+	{
+		logerror("write to palette_data2_w when not enabled %02x\n", data);
+	}
+	else
+	{
+		int offs = (m_palette_offset + (m_palette_offset_msb << 8)) & 0x1fff;
+		palette_data_lsb[offs] = data;
+		set_color(offs);
+	}
+}
+
+void anes_state::palette_data_msb_w(uint8_t data)
+{
+	if (m_palette_enable != 0x01)
+	{
+		logerror("write to palette_data_msb_w when not enabled %02x\n", data);
+	}
+	else
+	{
+		int offs = (m_palette_offset + (m_palette_offset_msb << 8)) & 0x1fff;
+		m_palette_data_msb[offs] = data;
+		set_color(offs);
+	}
+}
+
+
+void anes_state::opcodes_map(address_map &map)
+{
+	map(0x0000, 0xffff).r(FUNC(anes_state::m1_rom_r));
 }
 
 void anes_state::prg_map(address_map &map)
 {
-	map(0x0000, 0xefff).rom();
-	map(0xf000, 0xffff).ram();
+	map(0x0000, 0x1fff).rom();
+	map(0x2000, 0xefff).bankr(m_rombank);
+	map(0x0000, 0x01ff).w(FUNC(anes_state::blit_rom_w));
+	map(0xf000, 0xffff).ram(); // there might be a hole at 0xf800 - 0xf8ff
 }
 
 void anes_state::io_map(address_map &map)
 {
 	map.global_mask(0xff);
 	map.unmap_value_high();
 
-	map(0x07, 0x07).nopw(); // mux write
+	map(0x00, 0x0f).w(FUNC(anes_state::blit_w));
+
+	map(0x07, 0x07).w(FUNC(anes_state::matrix_w));
 	map(0x08, 0x09).w("ym", FUNC(ym2413_device::write));
-	map(0x0a, 0x0a).w(FUNC(anes_state::blit_trigger_w));
-//  map(0x0b, 0x0b).w(FUNC(anes_state::blit_mode_w));
-	map(0x0c, 0x0e).w(FUNC(anes_state::vram_offset_w));
-	map(0x11, 0x11).portr("DSW1");
-	map(0x12, 0x12).portr("DSW2");
-	map(0x13, 0x13).portr("DSW3");
-	map(0x14, 0x15).nopr(); // mux read
-	map(0x16, 0x16).portr("IN0").nopw();
-//  map(0xfe, 0xfe) banking? unknown ROM range
+
+	map(0x10, 0x10).portr("SW1");
+	map(0x11, 0x11).portr("SW2");
+	map(0x12, 0x12).portr("SW3");
+	map(0x13, 0x13).portr("SW4");
+	map(0x14, 0x15).r(FUNC(anes_state::key_r));
```

Comment:
more likely:
map(0x14, 0x14).r(FUNC(anes_state::key_r));
map(0x15, 0x15).portr("COIN");

COIN 0x40 = coin (other bits unknown)

Reply:
again, it's code I was sent, so I don't know why it's like that.